bot/database/__init__.py

```diff
@@ -0,0 +1,263 @@
+            await db.execute(sql, *sql_args)
+
+    async def db_fetchone(self, sql: str, sql_args: t.Optional[list] = None) -> asyncpg.Record:
+        """
+        This method serves as an abstraction layer
+        from using context manager and fetching the
+        sql query directly from there.
+        """
+        if not sql_args:
+            sql_args = []
+
+        async with self.pool.acquire(timeout=self.timeout) as db:
+            return await db.fetchrow(sql, *sql_args)
+
+    async def db_fetch(self, sql: str, sql_args: t.Optional[list] = None) -> t.List[asyncpg.Record]:
+        """
+        This method serves as an abstraction layer
+        from using context manager and fetching the
+        sql query directly from there.
+        """
+        if not sql_args:
+            sql_args = []
+
+        async with self.pool.acquire(timeout=self.timeout) as db:
+            return await db.fetch(sql, *sql_args)
+
+    async def db_get(self, column: t.List[str], specification: t.Optional[str] = None, sql_args: t.Optional[list] = None) -> asyncpg.Record:
+        """
+        This method serves as an abstraction layer
+        from using SQL syntax in the top-level database
+        table class, it runs the basic selection (get)
+        query without needing to use SQL syntax at all.
+        """
+        sql = f"SELECT {' ,'.join(column)} FROM {self.table}"
+        if specification:
+            sql += f" WHERE {specification}"
+
+        return await self.db_fetchone(sql, sql_args)
+
+    async def db_set(self, columns: t.List[str], values: t.List[str]) -> None:
+        """
+        This method serves as an abstraction layer
+        from using SQL syntax in the top-level database
+        table class, it runs the basic insertion (set)
+        command without needing to use SQL syntax at all.
+        """
+        sql_columns = ", ".join(columns)
+        sql_values = ", ".join(f"${n + 1}" for n in range(len(values)))
+
+        sql = f"""
+        INSERT INTO {self.table} ({sql_columns})
+        VALUES ({sql_values})
+        """
+
+        await self.db_execute(sql, values)
+
+    async def db_upsert(self, columns: t.List[str], values: t.List[str], conflict_column: str) -> None:
+        """
+        This method serves as an abstraction layer
+        from using SQL syntax in the top-level database
+        table class, it runs the basic insert/update (upsert)
+        command without needing to use SQL syntax at all.
+        """
+        sql_columns = ", ".join(columns)
+        sql_values = ", ".join(f"${n + 1}" for n in range(len(values)))
+        sql_update = ""
+        for index, column in enumerate(columns):
+            if column != conflict_column:
+                sql_update += f"{column}=${index + 1}"
+
+        sql = f"""
+        INSERT INTO {self.table} ({sql_columns})
+        VALUES ({sql_values})
+        ON CONFLICT ({conflict_column}) DO
+        UPDATE SET {sql_update}
+        """
```

Comment:
Why not directly do inside DB query, instead of Variables?

Reply:
I find it cleaner to do it this way but technically there wouldn't be anything wrong with putting it into the function directly if needed.